Call lift_fresh after scalar_to_tensor in composite derivative formulas

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81609

`scalar_to_tensor` is not dispatched and thus there is no interposition point for modes to ensure that the resulting tensor is appropriately wrapped. `lift_fresh` introduces this interposition point. This prevents FakeTensorMode from erroring. I can't make these wrapped numbers because there is some downstream logic on convolution backwards that expects these inputs to be honest to goodness tensors for conjugation.

This fixes test_aot_autograd_exhaustive_special_ndtr_cpu_float32
in https://github.com/pytorch/functorch/pull/935

See https://github.com/pytorch/pytorch/issues/81608 for more discussion

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Preserve wrapped numberness of Scalar overloads in derivative formulas

Without calling wrapped_scalar_tensor, the Scalar gets represented
as a regular tensor rather than a wrapped number, and if sent to
Python dispatch will show up as a Tensor, which is a problem if you
are running under something like FakeTensorMode which will choke.
Another possibility would have been to at::lift the resulting raw
tensor, but given that these really are scalars, preserving their
wrapped-numberness seemed better.

This fixes test_aot_autograd_exhaustive_special_ndtr_cpu_float32
in https://github.com/pytorch/functorch/pull/935

See https://github.com/pytorch/pytorch/issues/81608 for more discussion

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
Alright, this doesn't work, gonna try backup plan

ezyang committed:
Update on "Preserve wrapped numberness of Scalar overloads in derivative formulas"


Without calling wrapped_scalar_tensor, the Scalar gets represented
as a regular tensor rather than a wrapped number, and if sent to
Python dispatch will show up as a Tensor, which is a problem if you
are running under something like FakeTensorMode which will choke.
Another possibility would have been to at::lift the resulting raw
tensor, but given that these really are scalars, preserving their
wrapped-numberness seemed better.

This fixes test_aot_autograd_exhaustive_special_ndtr_cpu_float32
in https://github.com/pytorch/functorch/pull/935

See https://github.com/pytorch/pytorch/issues/81608 for more discussion

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Call lift_fresh after scalar_to_tensor in composite derivative formulas"


`scalar_to_tensor` is not dispatched and thus there is no interposition point for modes to ensure that the resulting tensor is appropriately wrapped. `lift_fresh` introduces this interposition point. This prevents FakeTensorMode from erroring. I can't make these wrapped numbers because there is some downstream logic on convolution backwards that expects these inputs to be honest to goodness tensors for conjugation.

This fixes test_aot_autograd_exhaustive_special_ndtr_cpu_float32
in https://github.com/pytorch/functorch/pull/935

See https://github.com/pytorch/pytorch/issues/81608 for more discussion

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Call lift_fresh after scalar_to_tensor in composite derivative formulas"


`scalar_to_tensor` is not dispatched and thus there is no interposition point for modes to ensure that the resulting tensor is appropriately wrapped. `lift_fresh` introduces this interposition point. This prevents FakeTensorMode from erroring. I can't make these wrapped numbers because there is some downstream logic on convolution backwards that expects these inputs to be honest to goodness tensors for conjugation.

This fixes test_aot_autograd_exhaustive_special_ndtr_cpu_float32
in https://github.com/pytorch/functorch/pull/935

See https://github.com/pytorch/pytorch/issues/81608 for more discussion

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -1678,7 +1678,12 @@
+  self: grad
+  result: auto_linear
+
+- name: lift_fresh(Tensor(a) self) -> Tensor(a)
```
`lift_fresh` only intended to be called when self is a Tensor materialized out of nowhere right? Wouldn't it always have requires_grad=False then?

ezyang replied:
yes... but when I ran the xla tests they errored because of a missing `lift_fresh` derivative, I guess I will need to investigate more

ezyang replied:
It's a matter of semantics. When a mode is applied, alias is expected to get tensor inputs that already have had all of the mode wrapping applied. `lift_fresh` is special: it is called specifically when (1) we know no mode wrapping could have been applied on the input (because we disabled all modes when allocating it) and (2) we know that the input tensor was freshly allocated (so it is sound to return an alias and mutate it.)

ezyang committed:
Update on "Call lift_fresh after scalar_to_tensor in composite derivative formulas"


`scalar_to_tensor` is not dispatched and thus there is no interposition point for modes to ensure that the resulting tensor is appropriately wrapped. `lift_fresh` introduces this interposition point. This prevents FakeTensorMode from erroring. I can't make these wrapped numbers because there is some downstream logic on convolution backwards that expects these inputs to be honest to goodness tensors for conjugation.

This fixes test_aot_autograd_exhaustive_special_ndtr_cpu_float32
in https://github.com/pytorch/functorch/pull/935

See https://github.com/pytorch/pytorch/issues/81608 for more discussion

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -g

jeanschmidt commented:
@pytorchbot revert -m "breaking internal builds" -c "ghfirst"

ezyang commented:
@pytorchbot rebase
